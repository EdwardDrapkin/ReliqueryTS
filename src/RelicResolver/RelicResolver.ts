import { InjectableClassification, InterfaceDescriptor, NamedFileItem } from "Types";
import { ResolutionError } from "ResolutionError";


interface LookupTableItem {
    registrations: NamedFileItem[]
}


export class RelicResolver {
    private lookupTable = new Map<string, LookupTableItem>();

    private static toIdentifierString(item: NamedFileItem) {
        return `${item.name}`;
    }

    private static getEmptyTableItem() {
        return {
            registrations: []
        }
    }

    private ensureTableItem(id: string) {
        if (!this.lookupTable.has(id)) {
            this.lookupTable.set(id, RelicResolver.getEmptyTableItem());
        }
    }

    private attachClassification(attach: InjectableClassification, to: NamedFileItem) {
        const interfaceId = RelicResolver.toIdentifierString(to);
        this.ensureTableItem(interfaceId);
        const { registrations } = this.lookupTable.get(interfaceId)!;
        const registered = registrations.find((item) => item.name === attach.name && item.filePath === attach.filePath);

        if (!registered) {
            const { name, filePath } = attach;

            registrations.push({ name, filePath });
        }
    }

    register(classification: InjectableClassification) {
        const classId = RelicResolver.toIdentifierString(classification);

        this.ensureTableItem(classId);

        this.attachClassification(classification, classification);

        classification.interfaces.forEach(interfaceHierarchy => {
            const attach = (item: InterfaceDescriptor) => this.attachClassification(classification, item);
            attach(interfaceHierarchy.implemented);
            interfaceHierarchy.parents.forEach(attach);
        });
    }

    lookup(id: string) {
        const entry = this.lookupTable.get(id);

        if (!entry) {
            throw new ResolutionError(`Could not resolve ${id}.`);
        }

        const { registrations } = entry;

        if (registrations.length < 0) {
            throw new ResolutionError(`Could not resolve ${id}.`);
        } else if (registrations.length > 1) {
            throw new ResolutionError(`Multiple resolutions for ${id}.`);
        }

        const { filePath, name } = registrations[0];
        return { filePath, name };
    }

    serialize() {
        return JSON.stringify([...this.lookupTable], null, 2);
    }

    deserialize(input: string) {
        this.lookupTable = new Map(JSON.parse(input));
    }
}
