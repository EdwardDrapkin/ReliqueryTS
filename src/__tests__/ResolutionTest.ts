import { RelicResolver } from "RelicResolver/RelicResolver";

describe("RelicResolver", () => {
    const resolver = new RelicResolver();

    test("A concrete class with no interfaces can be resolved", () => {
        resolver.register({
            filePath: "/fake/Fake.ts",
            name: "Fake",
            interfaces: []
        });

        const lookupClass = { filePath: "/fake/Fake.ts", name: "Fake" };
        expect(resolver.lookup(lookupClass)).toEqual(lookupClass);
        expect(resolver.serialize()).toMatchSnapshot();
    });

    test("A concrete class with one flat interface can be resolved", () => {
        resolver.register({
            filePath: "/fake/Fake2.ts",
            name: "Fake2",
            interfaces: [{
                implemented: {
                    name: "FakeInterface",
                    filePath: "/fake/FakeInterface.ts"
                },
                parents: []
            }]
        });

        const lookupClass = { filePath: "/fake/Fake2.ts", name: "Fake2" };
        expect(resolver.lookup(lookupClass)).toEqual(lookupClass);
        expect(resolver.serialize()).toMatchSnapshot();
    });

    test("A concrete class with multiple hierarchical interfaces can be resolved", () => {
        const classification = {
            filePath: "/fake/Fake3.ts",
            name: "Fake3",
            interfaces: [
                {
                    implemented: {
                        name: "FakeInterface2",
                        filePath: "/fake/FakeInterface2.ts"
                    },
                    parents: [{
                        name: "FakeInterface2Parent",
                        filePath: "/fake/FakeInterface2Parent.ts"
                    }, {
                        name: "FakeInterface2GrandParent",
                        filePath: "/fake/FakeInterface2GrandParent.ts"
                    }, {
                        name: "FakeInterface2GreatGrandParent",
                        filePath: "/fake/FakeInterface2GreatGrandParent.ts"
                    }]
                },
                {
                    implemented: {
                        name: "FakeInterface2a",
                        filePath: "/fake/FakeInterface2a.ts"
                    },
                    parents: [{
                        name: "FakeInterface2aParent",
                        filePath: "/fake/FakeInterface2aParent.ts"
                    }, {
                        name: "FakeInterface2aGrandParent",
                        filePath: "/fake/FakeInterface2aGrandParent.ts"
                    }, {
                        name: "FakeInterface2aGreatGrandParent",
                        filePath: "/fake/FakeInterface2aGreatGrandParent.ts"
                    }]
                },
                {
                    implemented: {
                        name: "FakeInterface2b",
                        filePath: "/fake/FakeInterface2b.ts"
                    },
                    parents: [{
                        name: "FakeInterface2bParent",
                        filePath: "/fake/FakeInterface2bParent.ts"
                    }, {
                        name: "FakeInterface2bGrandParent",
                        filePath: "/fake/FakeInterface2bGrandParent.ts"
                    }, {
                        name: "FakeInterface2bGreatGrandParent",
                        filePath: "/fake/FakeInterface2bGreatGrandParent.ts"
                    }]
                }]
        };

        resolver.register(classification);

        const lookupClass = { filePath: "/fake/Fake3.ts", name: "Fake3" };
        expect(resolver.lookup(lookupClass)).toEqual(lookupClass);

        classification.interfaces.forEach(item => {
            expect(resolver.lookup(item.implemented)).toEqual(lookupClass);
            item.parents.forEach(parent => {
                expect(resolver.lookup(parent)).toEqual(lookupClass)
            });
        });
        console.log(resolver.serialize());
        expect(resolver.serialize()).toMatchSnapshot();
    });
});
