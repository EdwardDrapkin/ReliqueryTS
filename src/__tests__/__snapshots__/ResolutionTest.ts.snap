// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelicResolver A concrete class with multiple hierarchical interfaces can be resolved 1`] = `
"[
  [
    \\"/fake/Fake.ts::Fake\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake\\",
          \\"filePath\\": \\"/fake/Fake.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/Fake2.ts::Fake2\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake2\\",
          \\"filePath\\": \\"/fake/Fake2.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface.ts::FakeInterface\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake2\\",
          \\"filePath\\": \\"/fake/Fake2.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/Fake3.ts::Fake3\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2.ts::FakeInterface2\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2Parent.ts::FakeInterface2Parent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2GrandParent.ts::FakeInterface2GrandParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2GreatGrandParent.ts::FakeInterface2GreatGrandParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2a.ts::FakeInterface2a\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2aParent.ts::FakeInterface2aParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2aGrandParent.ts::FakeInterface2aGrandParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2aGreatGrandParent.ts::FakeInterface2aGreatGrandParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2b.ts::FakeInterface2b\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2bParent.ts::FakeInterface2bParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2bGrandParent.ts::FakeInterface2bGrandParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface2bGreatGrandParent.ts::FakeInterface2bGreatGrandParent\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake3\\",
          \\"filePath\\": \\"/fake/Fake3.ts\\"
        }
      ]
    }
  ]
]"
`;

exports[`RelicResolver A concrete class with no interfaces can be resolved 1`] = `
"[
  [
    \\"/fake/Fake.ts::Fake\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake\\",
          \\"filePath\\": \\"/fake/Fake.ts\\"
        }
      ]
    }
  ]
]"
`;

exports[`RelicResolver A concrete class with one flat interface can be resolved 1`] = `
"[
  [
    \\"/fake/Fake.ts::Fake\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake\\",
          \\"filePath\\": \\"/fake/Fake.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/Fake2.ts::Fake2\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake2\\",
          \\"filePath\\": \\"/fake/Fake2.ts\\"
        }
      ]
    }
  ],
  [
    \\"/fake/FakeInterface.ts::FakeInterface\\",
    {
      \\"registrations\\": [
        {
          \\"name\\": \\"Fake2\\",
          \\"filePath\\": \\"/fake/Fake2.ts\\"
        }
      ]
    }
  ]
]"
`;
